### PR TITLE
Fixed #24638 -- Added support for SQL comments in queries

### PR DESCRIPTION
**Pull request by adamchainz**
[Ticket 24638](https://code.djangoproject.com/ticket/24638#ticket)

### CODE REVIEW

File: tests/test_mysql.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# This is an example test settings file for use with the Django test suite.
```
Comment (timgraham) [inline tests/test_mysql.py]: (please don't commit this file)

Comment (adamchainz) [inline tests/test_mysql.py]: Thanks for spotting!

**Comment by timgraham** [comment]
Is this ready for review from your perspective? "Needs documentation" is still checked on the ticket, hence it's not appearing in the review queue. It needs a rebase (might just be release notes).

**Comment by adamchainz** [comment]
I think it needs some more discussion. There hasn't been much enthusiasm on the ticket, and this is probably quite narrow in scope as using the comment for query hints would only really be work on MySQL.

And on that, the code I've published in my [django-mysql](https://github.com/adamchainz/django-mysql/blob/master/django_mysql/rewrite_query.py) package is slightly hacky (it works around the ORM by injecting comments with `extra()` and then rewriting them after the query is compiled) but sufficient for any use cases I've come across.

**Comment by timgraham** [comment]
Okay, please send a new PR if you revive it.

**Comment by thugdoubt** [comment]
I'd like to point out that the utility of this goes beyond just MySQL query hints. A prime example would be a postgresql pool- comments can be used to route queries, e.g. force a select to use the master. In addition, this would be an invaluable tool in tracing the source of slow queries.

**Comment by adamchainz** [comment]
@thugdoubt Where do the comments need to go though? One of the reasons I abandoned this is because it's too MySQL specific putting the comments directly after the `SELECT` keyword, since other DB's like their hints in other places. I suspect PostgreSQL pool routing has different rules again.

**Comment by Anton-Shutik** [comment]
@adamchainz 
>In PostgreSQL, a comment that starts with /* symbol and ends with */ and can be anywhere in your SQL statement. This method of commenting can span several lines within your SQL.

https://www.techonthenet.com/postgresql/comments.php

**Comment by adamchainz** [comment]
@Anton-Shutik yes, but this PR discusses special comments that the database parses for specific purposes.

**Comment by Anton-Shutik** [comment]
Oh, I see, you wanted to add some logic to sql query via comment. 
But adding comments on the query still might be useful, right ? We can pass there some meta data about server (or pod) that generated the query, which will be helpful when debugging.

**Comment by adamchainz** [comment]
Yes. If you want to do that, comments are quite easy to insert at the start of queries with [database instrumentation](https://docs.djangoproject.com/en/3.0/topics/db/instrumentation/) these days.

**Comment by Anton-Shutik** [comment]
Yeah, but let's say I want to add a comment to single sql query, [that](https://docs.djangoproject.com/en/3.0/topics/db/instrumentation/) approach looks like a hack where I have to edit generated sql, rather than a feature, right ? Approach suggested in the PR much nicer and addresses comments feature correctly.

**Comment by adamchainz** [comment]
It could maybe be done with a `.comment()` ORM method that has options on where to insert the comment, to allow the various use cases. More research is needed on all the backends.

**Comment by Anton-Shutik** [comment]
If I'm understanding correctly django ORM does not have `comment` method right now, and that is what this PR is adding, right ?

I've just walked through [supported](https://docs.djangoproject.com/en/3.0/ref/databases/) backends and looks like all of them support `/* comment goes here */` syntax, which might be inserted in any place of the query. [PostgreSQL](https://www.techonthenet.com/postgresql/comments.php), [MariaDB](https://www.techonthenet.com/mariadb/comments.php), [MySQL](https://www.techonthenet.com/mysql/comments.php), [Oracle](https://www.techonthenet.com/oracle/comments.php), [SQLite](https://www.techonthenet.com/sqlite/comments.php).

Considering that this feature is just comments addition I would just have it in front of the sql query  for now, and added "placement" options later, if required.

**Comment by adamchainz** [comment]
Yes you're right that both support it anywhere. For the MySQL/MariaDB query hints use case, the placement does matter.

This PR was closed 4 years ago. We'd need to start again on it. I'm afraid I don't have the time nor motivation to do so right now. Feel free to start a new PR - please check out the [contribution guide](https://docs.djangoproject.com/en/dev/internals/contributing/).

**Pull request closed without merging**